src/search/include/search/canonical_word.h

```diff
@@ -148,16 +142,205 @@ operator==(const ATARegionState<LocationT> &s1, const ATARegionState<LocationT>
+	// is partitioned by the fractional part and the region index can only be even if the fractional
+	// part is 0. If that is the case, there cannot be any configuration with an odd region index in
+	// the same partition, as that configuration's fractional part would be > 0.
+	std::for_each(word.begin(), word.end(), [&word](const auto &configurations) {
+		if (std::any_of(configurations.begin(),
+		                configurations.end(),
+		                [](const auto &w) { return get_region_index(w) % 2 == 0; })
+		    && std::any_of(configurations.begin(), configurations.end(), [](const auto &w) {
+			       return get_region_index(w) % 2 == 1;
+		       })) {
+			throw InvalidCanonicalWordException(word, "both odd and even region indexes");
+		}
+	});
+	// There must be at most one partition with fractional part 0.
+	// The only partition that is allowed to have fracitonal part 0 is the 0th
+	// partition.
+	std::for_each(std::next(word.begin()), word.end(), [&word](const auto &configurations) {
+		std::for_each(configurations.begin(), configurations.end(), [&word](const auto &w) {
+			if (get_region_index(w) % 2 == 0) {
+				throw InvalidCanonicalWordException(word,
+				                                    "fractional part 0 in wrong element of partition");
+			}
+		});
+	});
+	return true;
+}
+
+/** Get the canonical word H(s) for the given A/B configuration s, closely
+ * following Bouyer et al., 2006. The TAStates of s are first expanded into
+ * triples (location, clock, valuation) (one for each clock), and then merged
+ * with the pairs from the ATAConfiguration. The resulting set is then
+ * partitioned according to the fractional part of the clock valuations. Then,
+ * each tuple is regionalized by replacing the clock valuation with the
+ * respective region index.  The resulting word is a sequence of sets, each set
+ * containing regionalized tuples that describe a TAState or ATAState. The
+ * sequence is sorted by the fractional part of the original clock valuations.
+ * @param plant_configuration The configuration of the plant A (e.g., a TA
+ * configuration)
+ * @param ata_configuration The configuration of the alternating timed automaton B
+ * @param K The value of the largest constant any clock may be compared to
+ * @return The canonical word representing the state s, as a sorted vector of
+ * sets of tuples (triples from A and pairs from B).
+ */
+template <typename Location, typename ConstraintSymbolType>
+CanonicalABWord<Location, ConstraintSymbolType>
+get_canonical_word(const PlantConfiguration<Location> &          plant_configuration,
+                   const ATAConfiguration<ConstraintSymbolType> &ata_configuration,
+                   const unsigned int                            K)
+{
+	using ABSymbol       = ABSymbol<Location, ConstraintSymbolType>;
+	using ABRegionSymbol = ABRegionSymbol<Location, ConstraintSymbolType>;
+	// TODO Also accept a TA that does not have any clocks.
```

Comment:
Is this TODO still relevant? I think it is fair to assume that we have at least one clock, otherwise we are in a completely different synthesis setting, or not?

Reply:
Good catch! I think we should check all TODOs and clean them up if necessary, so I'll postpone this for now.